### ISSUE TITLE
Improve plugin upload/install capabilities

### ISSUE THREAD

**Issue by Bloke** [textpattern/textpattern#1489]
Would it be handy, in addition to being able to drag .zip and.php files to the Choose... button on the Admin->Plugins panel (or browse for them by clicking) that we accept .txt files there too? They could be detected and installed the same way into the DB / file system as if you'd copy n pasted them into the Upload box.

(If this was done we could potentially phase out the base64 pasteable Install textarea?)

I think it's pretty easy to do. Happy to take this on if people feel it'd be useful.

**Comment by bloatware** [comment]
That would be handy, but we'd need to pass by 'preview' step unlike for .zip and .php files. And Install textarea is still useful imo for copy/paste installs.

It would be interesting to be able to install plugins/themes from their URL too.

**Comment by Bloke** [comment]
Hmm, preview, yeah. Forgot about that!

Perhaps we skip preview if you browse for a file (php, zip or txt) but keep the preview step if you paste .txt content into the textarea? Gives people the option to see the code or not. I like the preview functionality as it allows me to grab the code and cancel the install so I can diff the code first.

Being a purist, we should preview all of the types, but that's a bit of a stretch I guess. Since we bypass preview for .zip and .php files, I don't think it's too much of an issue to bypass it for .txt files that are dragged in.

Thoughts?

**Comment by bloatware** [comment]
Dunno, .php are in clear text and .zip are bundles of multiple files, so previewing them has no much sense. But .txt are different and not everybody has a base64 decode reflex.

**Comment by Bloke** [comment]
True :)

We _could_ preview all. Simply unpack the contents and show the three main fields expanded as we do now. And for .zip files, list any additional files beneath the textpack field. Just so people know what's going to be uploaded/replaced.

That also opens up the possibility of adding checkboxes alongside each file in future. Defaults to all selected. But a nice boon if you've customised something and don't want a particular file or three overwritten in the archive.

Dunno.

**Comment by Bloke** [comment]
Also, if we're thinking of allowing install from url, you might not be able to read an archive or a .php file. So previewing all plugin content types would be a good security measure.

Not sure how we'd handle unpacking zip files without doing it twice.

**Comment by bloatware** [comment]
Feel free to experiment :-)

**Comment by Bloke** [comment]
Will do in a separate branch.

**Comment by Bloke** [comment]
Okay, modified the title of this topic. Here's the current thinking:

* Introduce preview step for _all_ plugin uploads. Zip/.php/.txt/base64 string, and those being imported from file via the dropdown.
* Permit .txt files to be installed via drop on the Plugin upload box.
* Permit URLs to be dragged into the Plugin upload box too (providing url_fopen is enabled).
* Split the `Plugin->read()` / `Plugin->install()` functionalty so that uploaded/imported/URL files are moved to a temporary filesystem area - Txp's tmp or system /tmp. i.e. they are NOT installed to the target PLUGINPATH at this stage.
* Add the temporary location path to the verify Form's payload.
* When extracting .zip or .php file contents, pull out 'data' entries as well as code/help/Textpacks and display the contents in a textarea too on the verify step.
* When extracting .zip files, pull out any additional filenames that are bundled and list those filenames beneath the textareas. So people know exactly what a plugin is going to do when it's installed.
* Optional/longer-term: offer checkboxes alongside everything except code / help / Textpack textareas. All checked by default. If any are unchecked by the admininstrator, the unchecked components are NOT installed when the 'Install' button is clicked.
* When clicking the Cancel button, do a proper form submission rather than a simple link back to the Plugin list panel. This allows temporary files to be removed if used. Using the browser 'back' button will leave the temp files behind but there's nothing much we can do about that. The files will just rot.
* Since files are in transit and cached in temp, and we're relying on base64 integrity not being compromised in the Verify form, make the Verify step construct an SHA1 hash of the incoming code and any additional files - a simple read of file contents hashed is fine. PHP has [hash_file](https://www.php.net/manual/en/function.hash-file.php) or [sha1_file](https://www.php.net/manual/en/function.sha1-file.php).

With regards the hash itself, store this in the txp_token table via a call to `generate_user_token()` where:

* `$ref` = numerical value derived from plugin name. Unique enough.
* `$type` = 'plugin_verify'
* `$expiryTimestamp` = maybe an hour or half hour from upload time?
* `$pass` = SHA1 of file/code contents
* `$nonce` = logged-in user nonce from the txp_user table - so if more than one person uploads the same plugin, they are distinguishable (temp file collisions notwithstanding - do we need to store the temp files with random names to circumvent this?)

Stash the returned token from `generate_user_token()` in the Verify form as a hidden field.

On submission with the 'Install' button:

* Retrieve the filename to the temp plugin file (if it exists) or base64 code.
* Recalculate the hash components from the incoming code parts.
* Retrieve the token based on the given hash selector and 'plugin_verify' type.
* Compare the timestamp and bounce if expired. Tidy up temp files if so.
* Compare the txp_token to the incoming hash from the verify form. If they don't match, file/code has been tampered with between upload/verify steps. Tidy up temp files if it doesn't match.
* If everything checks out, extract temporary files to their target destination in `PLUGINPATH` if necessary, install the nominated parts, add entry to DB, done. Tidy up temp files if necessary.

Once this is all in place, I'm confident we have a platform from which to launch the next phase: notification of updates on the Plugins panel based on data from the Plugin Repo site. And thus, simpler one-click 'Upgrade' installs directly from the Plugins panel that will fetch the code from the endpoint URL, download it, stick it in temp file location, unpack/generate token and show verification step. If Install is then clicked, perform upgrade from fetched file.

Ta-da!

**Comment by Bloke** [comment]
Phase one of the above is underway. It's great as I can now drag in a .txt, a .php or a .zip to the browse... button and Txp will preview them all and offer you the option to install or not, then just handle the corresponding type appropriately. There's one annoying thing though and I can't fathom what's going on.

@bloatware do you know what the rationale is behind 'writing' the plugin out to disk in the `Plugin->read()` method?

Specifically this:

``` php
            $plugin = array_filter(compact('code', 'help_raw'));

            if (!empty($code)) {
                file_put_contents($target_path, $pack);
                include $target_path;
            }
```
It seems if both a `name` and a `target_path` are passed in, the plugin does some jiggery pokery with extracting the code and help sections of the template file, then it writes some or all of it to disk in the PLUGINPATH directory. If you only pass in a name, then it presumes that the destination directory contains segregated components such as abc_plugin.php, manifest.json, help.textile and so forth.

When separating out this functionality to use a /tmp directory, which gives people the option of backing out after verifying, this 'forced write' to the PLUGINPATH is not appropriate.

So what I've done as a hack is always use the `($name, $target_path`) format. I simply pass in /tmp as the target_path when uploading, then pass in the real PLUGINPATH as target_path when the Install button is pressed, and pass in the location of the /tmp file as $name so it gets moved to the final destination.

The problem is that the `file_put_contents()` in the above code sees to be stripping out the code and help blocks and writing _some_ of the plugin template file as a .php file, then I dunno, maybe it's designed to then dump the code and help separately? This causes issues on the 'second pass' (install step, after verification) because the `Plugin->extractSection()` method returns nothing when parsing the file that has been stashed in /tmp by the first call to `->read()`.

Ideally I'd like to have it just divert the full file to the /tmp location at the 'verify' stage, then do any extraction only when 'install' is clicked. But it's kinda tricky and I'm trying to figure out what the purpose is of writing _most of_ the template file to disk minus the code and help blocks in the `->read()` method.

Any insight into why we do this please?

**Comment by bloatware** [comment]
@Bloke since all plugins are loaded from disk now, I thought it was faster to `include` only code, without associated (meta)data/help/etc. So each 'packed' plugin is split into its parts (`manifest.json` etc) on install. Why wouldn't you just move everything from `tmp` to `PLUGINPATH` on the second pass?

**Comment by Bloke** [comment]
> @Bloke since all plugins are loaded from disk now, I thought it was faster to `include` only code, without associated (meta)data/help/etc.

I agree.

> Why wouldn't you just move everything from `tmp` to `PLUGINPATH` on the second pass?

I can, but at the moment it's unpacking only the plugin metadata as abc_plugin.php file (without the code and help) and NOT the code or help as separate files. If I pass the _name only_ it works fine (albeit everything gets installed directly in PLUGINPATH). But if I pass in target_path, it doesn't do that. Been scratching my head trying to work out why.

Essentially, I want to refactor this a bit, but I'm trying not to break bwc. Having a method called `read()` which also (sometimes) writes to the filesystem is making matters a bit confusing, so I'm trying to sort out in my head what each bit does so I can rejig it a bit without changing the function signatures too much.

**Comment by bloatware** [comment]
I don't think there is any bwc risk to refactor this. I doubt it is (or will be) used anywhere outside core, save probably in Plugin Composer, but it needs to be adapted to disk storage anyway.

**Comment by Bloke** [comment]
True, Plugin Composer is super outdated now. Maybe I can just refactor core safely and do it in a way that allows us to choose the destination more easily to support the (now distinct) upload and verify steps. Will play, thanks.

**Comment by Bloke** [comment]
This is working well in my tests. Any issues or improvements, please note them here or go ahead and make them directly.

The next phase is to add some kind of hashing to protect uploaded plugins from tampering between verify and install stages. That shouldn't be too tricky, but placement of the code is going to be complicated by the fact that we won't have hash verification for plugin imports, yet it uses the same code to `Plugin->install()`. So the hash management might have to be performed from the panel code rather than the library (although the methods for doing so will probably be provided in the library, it won't use them itself).

Moving forward, I have absolutely no idea how to allow URLs to be dragged from one tab to another or to allow a URL to be pasted into the Browse widget and used to fetch plugin content directly from a server rather than from a local computer. I tried some JS trickery like @bloatware did in abc/com_article_image (that permits images to be dragged direcly from websites and inserted into an article) but it didn't seem to work. Just triggered a download of the dragged file each time.

It may simply be above my JavaScript pay grade so if anyone's willing to do that, by all means. An alternative might be to drag a link into the current Upload box. That's easy to differentiate from base64 and the URL could then be fetched via `file_get_contents()` (provided `allow_url_fopen` is enabled).

From the Plugins panel - once we get the repo up and running - an 'Update' link alongside any plugin could trigger a `file_get_contents()` from the nominated endpoint, and then we can divert to the verify step once it's retrieved to continue the installation process. So it may not be necessary to provide drag/drop from server environments, I only considered it because it might be handy for installing new plugins, as it removes the need to download it to the computer first then upload it to the site. Any thoughts/ideas on this welcome.

**Comment by bloatware** [comment]
Oh, that sounds mighty great! Will test more in details asap. Atm I get
```
Notice "Undefined index: code"
in C:\xampp\htdocs\textpattern\textpattern\include\txp_plugin.php at line 552.
```
on zip upload.

**Comment by Bloke** [comment]
Oooh, drat. Which plugin?

**Comment by philwareham** [comment]
In terms of auto update. We could either use the curated JSON repo for the data required, or the plugins website (each plugin page has JSON-LD content which gives the data required too).

**Comment by philwareham** [comment]
Basically I want to avoid linking directly to GitHub. The curated JSON list is (will be anyway) pulled into our plugins website at regular intervals.

**Comment by bloatware** [comment]
@Bloke whichever I tried (`abc_article_image` and `abc_prosemirror`)

**Comment by Bloke** [comment]
> Notice "Undefined index: code"
> in C:\xampp\htdocs\textpattern\textpattern\include\txp_plugin.php at line 552.

Hmmm, I think this might be a Windows/Linux thing as I can't reproduce it on my Linux box. I have a sneaky suspicion that inside a zip package, the files might be indexed with forward slashes _regardless_ of platform when brought into PHP. In which case, looking for a file that matches `{plugin_name}.DS.{plugin_name}.php` is going to fail.

Can you please confirm this by dumping out the contents of the `$zipFiles` array in `Plugin->read()`. If it reports the contents as abc_prosemirror/abc_prosemirror.php then we'll need to alter any references when matching files inside a zip to always use forward slashes, not DS.

**Comment by Bloke** [comment]
Actually, you know what, I'm placing bets on that being it so I've committed that change anyway. See c8e0e3fe.

**Comment by bloatware** [comment]
Bingo! Nice run.

**Comment by Bloke** [comment]
Plugins are now validated on install. Any tampering of the code or token between verify and install steps will result in a badly formed plugin error being thrown. Please try and hack in inventive ways to check this works.

Tokens are valid for one hour between verify and install. Is this enough? Too much for someone to manually verify that a plugin's code does what they need and isn't going to compromise their system? Tokens aren't yet tidied up from the DB: might do that in a later commit.

Note: new string required in the `[plugin]` group : `plugin_token_expired`. Will add this to the Textpack repo once we're sure this system works. We also require `plugin_install_cancelled` to be defined.

**Comment by Bloke** [comment]
I'm happy this is working okay, so please everyone, test it with as many plugins as you can, of any denomination: .txt/.php/.zip upload, or .txt paste.

@bloatware Do you think it's possible to permit direct opening of a URL so plugins stored on a server can be dragged directly to the Plugins panel upload form? If so, what avenues are open to us? Presumably it requires JS?

If not, that's cool. We can park development here until the plugin repo comes online, then carry on with Phase 3 of allowing updates directly from the Plugins panel.

**Comment by Bloke** [comment]
Plugin installation now tidies up after itself and removes tokens that are for the plugin just installed/cancelled, plus any old tokens that may have been left due to abandonment by using the browser's Back button/close tab or browser crash or whatever.

In the process, I introduced a new Security/Token class which handles all token-related activity. `form_token()` is now here (`Token->csrf()`) as is `bouncer()` (same name). We can gradually move over to using these methods as we go.

**Comment by bloatware** [comment]
@Bloke yes, I guess JS is required, but I've noticed that the properties of dragged objects can vary depending on the browser. I think some 'install from URL' field would be easier to handle.

**Comment by Bloke** [comment]
That's fine, thanks. Pasting/dragging a URL into the textarea could be a good way of doing that. Don't really want to add another bit of furniture to the page, and it would mean we keep the 'paste .txt box' around because it then serves a dual purpose. If it's a URL and url_fopen is enabled, fetch it and start the verify/install cycle depending on the type of content (zip, txt or php). Otherwise, see if it's BASE64 and continue as normal. Fail if whatever we get isn't of the expected format.

**Comment by bloatware** [comment]
Yep, and we could then try to redirect there URL drops on the file input if they can be detected reliably.

**Comment by Bloke** [comment]
388564a3 adds the ability to drag, type or paste a URL into the Install plugin textarea.

When you click Upload, the verification step checks if it's been handed a URL and, if so, calls the 'upload' step with the given URL so the contents can be fetched from the web and unpacked/stashed in the temp dir if necessary. The verify step is then automatically retriggered so the contents can be checked prior to installation. I'm hoping this doesn't end up in an infinite loop, so please test it with all manner of weird/mangled stuff and report back here.

See the commit message for additional caveats. Do we need some further checks here to test if the contents of any URL/uploaded file seems "reasonable?" Perhaps file size checks don't violate max upload size pref? Whether the content smells like its file extension indicates? Whether it actually contains PHP (if it's a PHP file)? That sort of thing.

@bloatware if you want to have a go at permitting URL drags onto the upload file input, be my guest. The JS required is above my pay grade. At the moment, `plugin_upload()` can only process URLs if they're passed into the function as the first param, but that could probably be tweaked/extended to perform the same action if the URL is dragged onto the file input box.

Note to self: we need to revisit the 4.9-branch pophelp for this panel when the dust settles as it's not accurate any more.

**Comment by bloatware** [comment]
@Bloke how would we handle dragging links onto the file input? Immediate download/verification or just copy `href` to the install textarea?

**Comment by Bloke** [comment]
Don't know what's best. If you drag a link in, I guess we have to auto-trigger the upload process, since we can't show '1 file' like we do with file upload boxes on the Images and Files panels....can we?

I prefer the idea of 'drag' and 'do it when I press the button' as separate actions, just for consistency with other panels and regular file inputs. But if that's not possible, auto-trigger of the 'upload' step which then pops up the verification step is fine. And it does save a click I guess.

Not convinced that dragging a file onto the Browse... area and then copying that href to the install area is very obvious UX. If you drag something to a file browser, you wouldn't expect to have to click a different Upload button to complete the action?

**Comment by bloatware** [comment]
Ok, let the dust settle, it's just few lines anyway.

**Comment by Bloke** [comment]
> In terms of auto update. We could either use the curated JSON repo for the data required, or the plugins website (each plugin page has JSON-LD content which gives the data required too).

Right. We need to decide on the most efficient way for a Textpattern site to query the plugin repo to see if there are updates available. Imagine someone has 30 plugins installed and visits their Admin->Plugins panel. We know the names of all of them, although not all may be displayed on Page 1 of the results. What do we do?

1. Go through each one and make a call to an endpoint that returns a machine-readable (JSON?) blob IF the plugin exists on our plugins site. When it comes back, compare the version number with the one installed and offer upgrade link if necessary. Repeat for each plugin.
2. Pass ALL plugin names as a single list to an endpoint on the plugins site. Get back a machine-readable structure (e.g. JSON blob) that contains the version info from all plugins that match the list. Less network traffic than option 1. Potentially larger payload.
3. Perform option 1 or option 2 but also pass in the _installed_ Textpattern version. The resulting JSON blob could then only have the list of plugins available that are compatible with the installed version. Possible additional UI furniture on the panel might be a checkbox that says 'ignore compatibility check' so people can get an entire set regardless of installed Txp version.
4. Perform option 1 or 2 and also pass each plugin's installed version so the check can be performed on the server, only returning plugins in the JSON blob that exceed the passed version number.
5. Something else?

In all of the above, perform a cache timeout (e.g. an hour) just like we do with the check for upgrades on the Diagnostics panel, so the plugin site isn't hammered for requests.

The Ajax nature of the panel muddies things a little because when you jump to page 2, or perform a search, we may already have the results from the plugin update check so we'd need to consult the results we already have (stashed in the prefs table as a hidden value?) and compare them to offer appropriate download links.

Anyone got any thoughts on best approach?

**Comment by Bloke** [comment]
Upgrades from the Plugins panel are now live. If a new version of a plugin is available in our curated plugins repository and verified compatible with your version of Txp:

* The 'version' string becomes clickable. It takes you to the plugin's page on plugins.textpattern.com (in a new browser tab).
* The Manage column gets upgrade links in it, for both stable and beta channels (if available).
* You can click the upgrade link to download the .txt/php/zip file and upload it via the Browse/Paste as normal.
* You can right-click the upgrade link and copy it, then paste it into the Upload box and hit 'Upload' to fetch the plugin directly from the repo.

Currently, there's the issue that very few of the plugins are verified compatible with 4.9.0 so they won't show up in the UI. I've updated a few manifests by hand for a few plugins I know work.

The upgrade check is performed when you visit the Plugins panel and is only performed a maximum of once an hour. It stashes the results in the Prefs table (in a hidden pref) and automatically refreshes this info when you install a plugin.

Please play.

**Comment by phiw13** [comment]
> Currently, there's the issue that very few of the plugins are verified compatible with 4.9.0 so they won't show up in the UI. I've updated a few manifests by hand for a few plugins I know work.

I was already wondering – how do  I test this? But then I noticed that a number of installed plugins indeed have a “plugin_upgrade” entry In the “Manage” column. 

But is this expected? 4 plugins have that “plugin_upgrade” entry, but all 4 point to the same plugin:
https://github.com/NicolasGraph/oui_cookie/releases/download/1.0.0-beta4/oui_cookie_v1.0.0-BETA4_zip.txt

The version string link is correct  and points to the plugin page on plugins.textpattern.com

**Comment by Bloke** [comment]
Phew, I was going to say, I swear I fixed that after I forgot to put the [resets in the right place](https://github.com/textpattern/textpattern-plugins-website/commit/33fef82e6cbb9f4bf9492e58bd9012095b83e63b).

It's probably a cache thing, but if it happens again as the community [populate more plugins](https://github.com/textpattern/textpattern-curated-plugins-list) then do holler.

**Comment by Bloke** [comment]
While the OP is technically already implemented, the recent changes to the plugins panel vis a vis the additional metadata fields means there is one final piece to make this battle ready: to permit compilation of the plugin from the edit panel. Got my thinking brain on for that one.

Compilation from multi-edit is more tricky, so I'm tempted to shelve that idea. But on the back burner is some investigation into how Gocom manages his Mass Plugin Compiler.

**Comment by petecooper** [comment]
>one final piece to make this battle ready: to permit compilation of the plugin from the edit panel

@Bloke Out of interest, how viable would this be for 4.9.0? Is it hideously complex?

**Comment by Bloke** [comment]
It's not technically complicated (plugin composer does half of it with a simple function). A few minor sticking points:

A) We need to either fold the [plugin template file](https://github.com/textpattern/textpattern-plugin-template/blob/main/zem_plugin.php) into core or fetch it on demand when compiling, which adds network overhead (what if it's not available, fallback, blah blah) and a dependency (we need to start versioning the template repo and making a static release each time we release a new Txp) or duplicate the template in core, which means the repo potentially falls out of sync with it.

B) Extracting textpack strings to go into the compiled plugin is currently done by either using an `owner` or a prefix, either of which would have to exactly match the plugin name. Not all plugins do this (e.g. smd_user_manager strings are all prefixed `smd_um`) so dragging textpack strings along for the ride requires some convention work on existing plugin authors, or on people who have installed their plugins and want to compile them.

C) This only covers standard resources. There's no provision for plugins to be bundled up with either additional `data` database content or extra files in its directory. The first would rely on someone injecting the content into the database via a DB admin tool like Sequel Ace, phpMyAdmin, etc and the compiler being told to check for it. The second is only doable via a .zip bundle after someone populates the plugin's on-disk directory with files: then the plugin export feature currently picks those up in the bundle.

All technically easy but until conventions are more widely adopted, the second two issues will be suck it and see.

Happy to have a go folding this in today.

**Comment by bloatware** [comment]
A) It's only few lines of code, folding it into core looks painless.

B) Are not help/strings/meta also stored on disk now? Why not to pick them from the corresponding files?

C) Caveat utilitor

**Comment by Bloke** [comment]
Ah yeah, hadn't thought about using the files on disk. That would indeed be simpler. Good idea. Is the `data` stuff stored there too, I can't remember?

I agree we should do this. I'll have a stab at it.

**Comment by bloatware** [comment]
Data should be in `data.txp`, but I have never tested it.

**Comment by Bloke** [comment]
Nice, thanks. I just checked though and we can use `Plugin>fetchData()` and if it's there, great, if not, skip it. Simple.

**Comment by Bloke** [comment]
It was actually even easier than expected because `Plugin->read()` already does everything for us. In theory, no need to faff with lang strings and data because it's all there for us on disk, and that function pulls it all into memory for us.

As noted in the [commit message](https://github.com/textpattern/textpattern/commit/208038a1ce81973ea268302cefbd19f1bf965448) there's no UI route to serve the .txt file uncompressed. It shouldn't ever be needed, but for legacy reasons the ability is preserved but it requires manual intervention by adding `&full=1` to the compile URL link. That's not the greatest verbTechnically it should be `&uncompress=1` or something like that, but that's more typing and I dislike negative logic.

Using `&zip={bool}` would be an option if we switched the URLs to always include that param, I guess. Was trying to minimise the number of params but it doesn't add much length. Thoughts?

**Comment by bloatware** [comment]
Wow, impressive! Is the final `_zip` in file name intentional?

**Comment by Bloke** [comment]
The _zip is only for legacy reasons. We can choose between base64 and gzipped base64, so this differentiates which is which. There have been instances in the past where some servers couldn't install gzipped plugins so the larger ones were used as a fallback. Happy to tweak the filename template now we have actual .zip files, to avoid confusion. Any preference on naming convention?

I think I will change it so that `&compress=1` is used all the time in the URL. It makes more sense and is then easier to flip between gzipped and non-gzipped versions.

**Comment by Bloke** [comment]
And let's use [compressed plugins by default](https://github.com/textpattern/textpattern/commits/dev/) which will only add the extra 'uncompressed' designation if you manually tinker with the URL. Cleaner and less confusing.

**Comment by petecooper** [comment]
@Bloke - what's the latest sit rep on this issue, please? It feels like it's pretty close to complete, but I may be missing something. Thank you.

**Comment by Bloke** [comment]
As far as I'm concerned, this is done. The stuff about selectively installing bits and pieces of bundled zips can wait.
